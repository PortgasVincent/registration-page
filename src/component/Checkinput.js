import React from 'react';

function Checkinput(props){
  return(
		<div className={"inputCheck"}>
			<label className={"inputLabel"}>{props.title}</label>
			<input type={props.type} onChange={props.checkformat} className={"inputText"} />
			<span className={props.isValid? "active" : "inactive"}>{props.format}</span>
		</div>
		)
  }
export {Checkinput};