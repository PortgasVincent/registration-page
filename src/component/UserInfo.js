import React from 'react';

function UserInfo(props){
	return(
		<div>
			<h3>User name:{props.name}</h3>
			<h3>Password:{props.psw}</h3>
			<h3>Phone:{props.phone}</h3>
		</div>
	)
}
export {UserInfo};