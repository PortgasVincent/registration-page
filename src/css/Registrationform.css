.title{
  text-align: center;
}
.regForm{
  width:100%;
}
.submitBtn{
  background-color:skyblue;
  width:300px;
  height:36px;
  margin-left:640px;
  font-size:20px;
  color:white;
}