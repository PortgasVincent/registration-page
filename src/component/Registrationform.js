import React from 'react';
import {Checkinput} from './Checkinput';
import {Userinfo} from './Userinfo';

class Registrationform extends React.Component{
  constructor(props){
    super();
    this.timer = null;
    this.data = {
      name:"",
      psw:"",
      confirmpsw:"",
      phone:"",
    };
    this.state={
      permittedname:false,
      permittedpsw:false,
      confirmedpsw:false,
      permittedphone:false,
      submitted:false,
    }
    this.handleSubmit = this.handleSubmit.bind(this);
    this.checkName = this.checkName.bind(this);
    this.checkPsw = this.checkPsw.bind(this);
    this.confirmPsw = this.confirmPsw.bind(this);
    this.checkPhone = this.checkPhone.bind(this);
  }
  checkName(e){
    const namestr = e.target.value;
    clearTimeout(this.timer);
    this.timer = setTimeout((namestr) => {
      const name = this.delHtmlTag( namestr );
      this.data.name = name;
      this.setState((prestate)=>{
        prestate.permittedname = name !== "";
        prestate.submitted = false;
        return prestate;
      })
    } ,1000 ,namestr);
		
  }
  checkPsw(e){
    const pswstr = e.target.value;
    clearTimeout(this.timer);
    this.timer = setTimeout((pswstr)=>{
      const psw = this.delHtmlTag( pswstr );
      const regUpper = /[A-Z]/;
      const reglower = /[a-z]/;
      this.data.psw = psw;
      this.setState((prestate)=>{
        prestate.permittedpsw = regUpper.test(psw) && reglower.test(psw) && psw.length >= 8;
        prestate.submitted = false;
        prestate.confirmedpsw = psw === this.data.confirmpsw;
        return prestate;
      })
    } ,1000 ,pswstr);
  }
  confirmPsw(e){
    const pswstr = e.target.value;
    clearTimeout(this.timer);
    this.timer = setTimeout((pswstr)=>{
      const psw = this.delHtmlTag( pswstr);
      this.data.confirmpsw = psw;
      this.setState((prestate)=>{
        prestate.submitted = false;
        prestate.confirmedpsw = psw === this.data.psw;
        return prestate;
      })
    } ,1000 ,pswstr);
  }
  checkPhone(e){
    const phonestr = e.target.value;
    clearTimeout(this.timer);
    this.timer = setTimeout((phonestr)=>{
      const phone = this.delHtmlTag( phonestr );
      this.data.phone = phone;
      this.setState((prestate)=>{
        prestate.permittedphone = (/^1(3|4|5|7|8)\d{9}$/).test(phone) || (/^(\(\d{3,4}\)|\d{3,4}-|\s)?\d{7,14}$/).test(phone);
        prestate.submitted = false;
        return prestate;
      })
    } ,1000 ,phonestr);
  }

  handleSubmit(e){
    this.setState({
      submitted: true,
    })
    e.preventDefault();
  }

  delHtmlTag(str)
  {
    let result=str.replace(/<\/?[^>]*>/gim,"");//delete html tag
    result=str.replace(/(^\s+)|(\s+$)/g,"");//delete space in begin or end
    return  result.replace(/\s/g,"");//delete space in center
  }
  render(){
    return(
      <form onSubmit={this.handleSubmit} className={"regForm"}>
        <Checkinput type={"text"}     title={"User name:"}        checkformat={this.checkName}  isValid={this.state.permittedname}  format={"Must not be null"}  />
        <Checkinput type={"password"} title={"Password:"}          checkformat={this.checkPsw}   isValid={this.state.permittedpsw}   format={"must be at least 8 characters, contain upper and lower case letters"}  />
        <Checkinput type={"password"} title={"Confirm Password:"} checkformat={this.confirmPsw} isValid={this.state.confirmedpsw}     format={"must match above"} />
        <Checkinput type={"text"}     title={"Phone Number:"}     checkformat={this.checkPhone} isValid={this.state.permittedphone} format={"ensure the number is valid for a Chinese mobile or fixed line phone number"}   />
        <input type="submit" value="Submit" disabled={!(this.state.permittedname && this.state.permittedpsw && this.state.confirmedpsw && this.state.permittedphone)} className={"submitBtn"} />
        {this.state.submitted && <Userinfo name={this.data.name} psw={this.data.psw} phone={this.data.phone} />}
      </form>
    );
  }
}
export {Registrationform};
