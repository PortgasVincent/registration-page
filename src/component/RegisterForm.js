import React from 'react';
import {Checkinput} from './Checkinput';
import {UserInfo} from './UserInfo';

class RegistrationForm extends React.Component{
	timer;
	constructor(props){
		super();
		this.state={
			data:{
				name:"",
				psw:"",
				confirmpsw:"",
				phone:"",
			},	
			permittedname:false,
			permittedpsw:false,
			confirmedpsw:false,
			permittedphone:false,
			submitted:false,
		}
		// this.timer = "";
		this.handleSubmit = this.handleSubmit.bind(this);
		this.checkName = this.checkName.bind(this);
		this.checkPsw = this.checkPsw.bind(this);
		this.confirmPsw = this.confirmPsw.bind(this);
		this.checkPhone = this.checkPhone.bind(this);
	}
	checkName(e){
		const namestr = e.target.value;
		clearTimeout(this.timer);
		this.timer = setTimeout((namestr) => {
			const name = this.delHtmlTag( namestr );
			if(name !== ""){
				this.setState((prestate)=>{
					prestate.data.name = name;
					prestate.permittedname = true;
					prestate.submitted = false;
					return prestate;
				})
			}else{
				this.setState((prestate)=>{
					prestate.data.name = name;
					prestate.permittedname = false;
					prestate.submitted = false;
					return prestate;
				})
			}
		} ,1000 ,namestr);
		
	}
	checkPsw(e){
		const pswstr = e.target.value;
		clearTimeout(this.timer);
		this.timer = setTimeout((pswstr)=>{
			const psw = this.delHtmlTag( pswstr );
			const regUpper = /[A-Z]/;
			const reglower = /[a-z]/;
			if(regUpper.test(psw) && reglower.test(psw) && psw.length >= 8){
				this.setState((prestate)=>{
					prestate.data.psw = psw;
					prestate.permittedpsw = true;
					prestate.submitted = false;
					return prestate;
				})
			}else{
				this.setState((prestate)=>{
					prestate.data.psw = psw;
					prestate.permittedpsw = false;
					prestate.submitted = false;
					return prestate;
				})
			}
			if(psw === this.state.data.confirmpsw){
				this.setState({
					confirmedpsw:true,
				})
			}else{
				this.setState({
					confirmedpsw:false,
				})
			}
		} ,1000 ,pswstr);
	}
	confirmPsw(e){
		const pswstr = e.target.value;
		clearTimeout(this.timer);
		this.timer = setTimeout((pswstr)=>{
			const psw = this.delHtmlTag( pswstr);
			this.setState((prestate)=>{
					prestate.data.confirmpsw = psw;
					prestate.submitted = false;
					return prestate;
				})
			if(psw === this.state.data.psw){
				this.setState({
					confirmedpsw: true,
				})
			}else{
				this.setState({
					confirmedpsw:false,
				})
			}
		} ,1000 ,pswstr);
	}
	checkPhone(e){
		const phonestr = e.target.value;
		clearTimeout(this.timer);
		this.timer = setTimeout((phonestr)=>{
			const phone = this.delHtmlTag( phonestr );
			if(phone.match(/^1(3|4|5|7|8)\d{9}$/) || phone.match(/^(\(\d{3,4}\)|\d{3,4}-|\s)?\d{7,14}$/)){
				this.setState((prestate)=>{
					prestate.data.phone = phone;
					prestate.permittedphone = true;
					prestate.submitted = false;
					return prestate;
				})
			}else{
				this.setState((prestate)=>{
					prestate.data.phone = phone;
					prestate.permittedphone = false;
					prestate.submitted = false;
					return prestate;
				})
			}
			//console.log(this.state.phone);
			//console.log(this.state.permittedphone);
		} ,1000 ,phonestr);
	}

	handleSubmit(e){
		this.setState({
			submitted: true,
		})
		const dataJson = {
			name:this.state.data.name, 
			psw:this.state.data.psw, 
			phone:this.state.data.phone,
		};
		e.preventDefault();
	}

	delHtmlTag(str)
	{
    let result=str.replace(/<\/?[^>]*>/gim,"");//delete html tag
    result=str.replace(/(^\s+)|(\s+$)/g,"");//delete space in begin or end
    return  result.replace(/\s/g,"");//delete space in center
	}
	render(){
		return(
			<form onSubmit={this.handleSubmit} className={"regForm"}>
				<Checkinput type={"text"}     title={"User name:"}        checkformat={this.checkName}  isValid={this.state.permittedname}  format={"Must not be null"}  />
				<Checkinput type={"password"} title={"Password"}          checkformat={this.checkPsw}   isValid={this.state.permittedpsw}   format={"must be at least 8 characters, contain upper and lower case letters"}  />
				<Checkinput type={"password"} title={"Confirm Password:"} checkformat={this.confirmPsw} isValid={this.state.confirmedpsw}     format={"must match above"} />
				<Checkinput type={"text"}     title={"Phone Number:"}     checkformat={this.checkPhone} isValid={this.state.permittedphone} format={"ensure the number is valid for a Chinese mobile or fixed line phone number"}   />
				<input type="submit" value="Submit" disabled={!(this.state.permittedname && this.state.permittedpsw && this.state.confirmedpsw && this.state.permittedphone)} className={"submitBtn"} />
				{this.state.submitted?
					<UserInfo name={this.state.data.name} psw={this.state.data.psw} phone={this.state.data.phone} /> : ""
				}
			</form>
		);
	}
}
export {RegistrationForm};
