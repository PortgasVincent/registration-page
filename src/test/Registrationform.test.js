import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {mount,shallow} from 'enzyme';
import {Registrationform} from '../component/Registrationform';
Enzyme.configure({ adapter: new Adapter() });

jest.useFakeTimers();

it("form", ()=>{
  const form = mount(<RegistrationForm />);
  //check init state
  expect(form.state().permittedname).toBe(false);
  expect(form.state().permittedpsw).toBe(false);
  expect(form.state().confirmedpsw).toBe(false);
  expect(form.state().permittedphone).toBe(false);
  expect(form.state().submitted).toBe(false);
  //get input
  const nameinput = form.find({title : "User name:"}).find("input");
  const pswinput = form.find({title : "Password"}).find("input");
  const cpswinput = form.find({title : "Confirm Password:"}).find("input");
  const phoneinput = form.find({title : "Phone Number:"}).find("input");
  const submitbtn = form.find(".submitBtn");
  //simulate onchange  simulate can run multiple events at the same time    
  //but the setTimeout is load in this.timer    so if you set new, it will replace 

  ////unavailable input
  //check name
  nameinput.simulate("change", {target:{value:" "}});
  //run inner of timer   simulate can run multiple events at the same time    
  jest.runAllTimers();
  //check value of input 
  expect(form.state().data.name).toBe("");
  //check value of input whether is permitted
  expect(form.state().permittedname).toBe(false);
  // check psw
  pswinput.simulate("change", {target:{value:"Aaaaaa"}});
  jest.runAllTimers();
  expect(form.state().data.psw).toBe("Aaaaaa");
  expect(form.state().permittedpsw).not.toBe(true);
  pswinput.simulate("change", {target:{value:"aaaaaaaa"}});
  jest.runAllTimers();
  expect(form.state().data.psw).toBe("aaaaaaaa");
  expect(form.state().permittedpsw).not.toBe(true);
  //check confirm psw
  cpswinput.simulate("change", {target:{value:"Bbbbbb"}});
  jest.runAllTimers();
  expect(form.state().data.confirmpsw).toBe("Bbbbbb");
  expect(form.state().confirmedpsw).not.toBe(true);
  //check phone
  phoneinput.simulate("change", {target:{value:"123456"}});
  jest.runAllTimers();
  expect(form.state().data.phone).toBe("123456");
  expect(form.state().permittedphone).not.toBe(true);
  expect(form.html()).toMatchSnapshot();

  //available input
  //check name
  nameinput.simulate("change", {target:{value:"aaa"}});
  //run inner of timer
  jest.runAllTimers();
  //check value of input 
  expect(form.state().data.name).toBe("aaa");
  //check value of input whether is permitted
  expect(form.state().permittedname).toBe(true);
  // check psw
  pswinput.simulate("change", {target:{value:"Aaaaaaaa"}});
  jest.runAllTimers();
  expect(form.state().data.psw).toBe("Aaaaaaaa");
  expect(form.state().permittedpsw).toBe(true);
  //check confirm psw
  cpswinput.simulate("change", {target:{value:"Aaaaaaaa"}});
  jest.runAllTimers();
  expect(form.state().data.confirmpsw).toBe("Aaaaaaaa");
  expect(form.state().confirmedpsw).toBe(true);
  //check phone
  phoneinput.simulate("change", {target:{value:"82408888"}});
  jest.runAllTimers();
  expect(form.state().data.phone).toBe("82408888");
  expect(form.state().permittedphone).toBe(true);
  //submit
  submitbtn.simulate("submit");
  expect(form.state().submitted).toBe(true);
  expect(form.html()).toMatchSnapshot();
});