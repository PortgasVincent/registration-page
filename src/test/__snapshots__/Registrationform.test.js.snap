// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form 1`] = `"<form class=\\"regForm\\"><div class=\\"inputCheck\\"><label class=\\"inputLabel\\">User name:</label><input type=\\"text\\" class=\\"inputText\\"><span class=\\"inactive\\">Must not be null</span></div><div class=\\"inputCheck\\"><label class=\\"inputLabel\\">Password:</label><input type=\\"password\\" class=\\"inputText\\"><span class=\\"inactive\\">must be at least 8 characters, contain upper and lower case letters</span></div><div class=\\"inputCheck\\"><label class=\\"inputLabel\\">Confirm Password:</label><input type=\\"password\\" class=\\"inputText\\"><span class=\\"active\\">must match above</span></div><div class=\\"inputCheck\\"><label class=\\"inputLabel\\">Phone Number:</label><input type=\\"text\\" class=\\"inputText\\"><span class=\\"inactive\\">ensure the number is valid for a Chinese mobile or fixed line phone number</span></div><input type=\\"submit\\" value=\\"Submit\\" disabled=\\"\\" class=\\"submitBtn\\"></form>"`;

exports[`form 2`] = `"<form class=\\"regForm\\"><div class=\\"inputCheck\\"><label class=\\"inputLabel\\">User name:</label><input type=\\"text\\" class=\\"inputText\\"><span class=\\"active\\">Must not be null</span></div><div class=\\"inputCheck\\"><label class=\\"inputLabel\\">Password:</label><input type=\\"password\\" class=\\"inputText\\"><span class=\\"active\\">must be at least 8 characters, contain upper and lower case letters</span></div><div class=\\"inputCheck\\"><label class=\\"inputLabel\\">Confirm Password:</label><input type=\\"password\\" class=\\"inputText\\"><span class=\\"active\\">must match above</span></div><div class=\\"inputCheck\\"><label class=\\"inputLabel\\">Phone Number:</label><input type=\\"text\\" class=\\"inputText\\"><span class=\\"active\\">ensure the number is valid for a Chinese mobile or fixed line phone number</span></div><input type=\\"submit\\" value=\\"Submit\\" class=\\"submitBtn\\"><div><h3>User name:aaa</h3><h3>Password:Aaaaaaaa</h3><h3>Phone:82408888</h3></div></form>"`;
